en/mavgen_c/message_signing_c.md

```diff
@@ -0,0 +1,149 @@
+# C Message Signing (mavgen)
+
+One of the key features of [MAVLink 2](../guide/mavlink_2.md) is support for [Message Signing](../guide/message_signing.md) (authentication).
+
+The C libraries generated by *mavgen* provide almost everything needed to support signing in your MAVLink system.
+You will need to add some code to:
+
+* Handle the [SETUP_SIGNING](../messages/common.md#SETUP_SIGNING) message.
+* Setup and teardown signing on a link.
+* Save and load the secret key and timestamp in persistent storage
+* Implement a callback to define which (if any) unsigned messages will be accepted.
+
+
+## Secret Key Management (SETUP_SIGNING)
+
+A secret key is 32 bytes of binary data that are used to create message signatures that can be verified by other holders of the key.
+The general requirements for creating, storing, logging and sharing keys are covered in: [Message Signing > Secret Key Management](../guide/message_signing.md#secret_key).
+
+<!-- 
+The [SETUP_SIGNING](../messages/common.md#SETUP_SIGNING) message should generally be used for sharing the secret key, and support for it must be implemented on both sending and receiving systems. Receiving systems must also store the key in secure storage. 
+
+how pass key to system after calculation? i.e. is there a set-key method? 
+What this should show is 
+- how to generate sha256 from paraphrase
+- how to handle received message and store the key (on nuttx and Linux)
+-->
+
+
+## Handling Timestamps
+
+The timestamp is a 48 bit number with units of 10 microseconds since 1st January 2015 GMT. 
+The general requirements for managing timestamps are covered in [Message Signing > Timestamp Handling](../guide/message_signing.md#timestamp).
+
+The library automatically handles some of the rules: 
+- timestamps are incremented by one on every message send from a link.
+- timestamps are updated to match that of the last accepted incoming message (if it is greater than the current local timestamp).
+- messages are rejected if the timestamp of a message on a channel is before the last message received on that channel.
+
+It is the responsibility of each MAVLink system to store and restore the timestamp into persistent storage (this is critical for the security of the signing system).
+
```

Comment:
SO what is missing here is code examples showing how to:
1. Get the current timestamp from library (for storing). 
2. Set the current timestamp in the library (e.g. to call when GPS gets a lock).
3. Set the timestamp that the library should use when first started.

It might also be cool if the library gave us a callback with the timestamp every minute for saving.

Reply:
Ah, I see "Enabling Signing on a Channel" shows how to get/set a timestamp for a channel.